dash-pipeline/dpapp/dash/node.c

```diff
@@ -0,0 +1,354 @@
+#undef _
+#define _(a) en0->dst_address[a] = tmp0[a];
+	  foreach_mac_address_offset;
+#undef _
+
+	  en1 = (ethernet_header_t*)dh1 - 1;
+#define _(a) tmp1[a] = en1->src_address[a];
+	  foreach_mac_address_offset;
+#undef _
+#define _(a) en1->src_address[a] = en1->dst_address[a];
+	  foreach_mac_address_offset;
+#undef _
+#define _(a) en1->dst_address[a] = tmp1[a];
+	  foreach_mac_address_offset;
+#undef _
+
+	  /* Update dash header */
+
+	  dh0->packet_meta.packet_source = DPAPP;
+	  if (dh0->packet_meta.packet_subtype != FLOW_DELETE) {
+	    /* Only keep packet_meta and flow_key in dash_header_t */
+	    u16 length0 = ntohs(dh0->packet_meta.length);
+	    dh0->packet_meta.length = htons(offsetof(dash_header_t, flow_data));
+	    /* Move customer packet after dash header */
+	    clib_memmove((u8*)&dh0->flow_data, (u8*)dh0 + length0, vlib_buffer_get_tail(b0) - (u8*)dh0 - length0);
+	    b0->current_length -= length0 - offsetof(dash_header_t, flow_data);
+	  }
+
+	  dh1->packet_meta.packet_source = DPAPP;
+	  if (dh1->packet_meta.packet_subtype != FLOW_DELETE) {
+	    /* Only keep packet_meta and flow_key in dash_header_t */
+	    u16 length1 = ntohs(dh1->packet_meta.length);
+	    dh1->packet_meta.length = htons(offsetof(dash_header_t, flow_data));
+	    /* Move customer packet after dash header */
+	    clib_memmove((u8*)&dh1->flow_data, (u8*)dh1 + length1, vlib_buffer_get_tail(b1) - (u8*)dh1 - length1);
+	    b1->current_length -= length1 - offsetof(dash_header_t, flow_data);
+	  }
+
+	  vlib_buffer_reset (b0);
+	  vlib_buffer_reset (b1);
+
+
+	  sw_if_index0 = vnet_buffer (b0)->sw_if_index[VLIB_RX];
+	  sw_if_index1 = vnet_buffer (b1)->sw_if_index[VLIB_RX];
+
+	  /* Send pkt back out the RX interface */
+	  vnet_buffer (b0)->sw_if_index[VLIB_TX] = sw_if_index0;
+	  vnet_buffer (b1)->sw_if_index[VLIB_TX] = sw_if_index1;
+
+	  pkts_swapped += 2;
+
+	  if (PREDICT_FALSE ((node->flags & VLIB_NODE_FLAG_TRACE)))
+	    {
+	      if (b0->flags & VLIB_BUFFER_IS_TRACED)
+		{
+		  dash_trace_t *t =
+		    vlib_add_trace (vm, node, b0, sizeof (*t));
+		  t->sw_if_index = sw_if_index0;
+		  t->next_index = next0;
+		  clib_memcpy_fast (t->new_src_mac, en0->src_address,
+				    sizeof (t->new_src_mac));
+		  clib_memcpy_fast (t->new_dst_mac, en0->dst_address,
+				    sizeof (t->new_dst_mac));
+
+		}
+	      if (b1->flags & VLIB_BUFFER_IS_TRACED)
+		{
+		  dash_trace_t *t =
+		    vlib_add_trace (vm, node, b1, sizeof (*t));
+		  t->sw_if_index = sw_if_index1;
+		  t->next_index = next1;
+		  clib_memcpy_fast (t->new_src_mac, en1->src_address,
+				    sizeof (t->new_src_mac));
+		  clib_memcpy_fast (t->new_dst_mac, en1->dst_address,
+				    sizeof (t->new_dst_mac));
+		}
+	    }
+
+	  /* verify speculative enqueues, maybe switch current next frame */
+	  vlib_validate_buffer_enqueue_x2 (vm, node, next_index,
+					   to_next, n_left_to_next,
+					   bi0, bi1, next0, next1);
+	}
+
+      while (n_left_from > 0 && n_left_to_next > 0)
+	{
+	  u32 bi0;
```

Comment:
probably can get rid of it to make the code more clear and maintainable. this huge chunk of duplicated code is not really ideal.

Reply:
fixed